[Draft PR] Support cgroups v1 and v2 based on installed LV version

pratheekshasn opened the pull request:
### Summary of Changes

This pull request addresses the handling of cgroups v1 and v2 on NILRT targets, and the version is chosen based on the version of LabVIEW Real Time installed on it. If the LV version is 2023 and below, we choose cgroups v1, if not, we use cgroups v2.

This requires a change to the docker-moby recipe, where I need to remove dependency to cgroup-lite. This is because cgroup-lite is specific to cgroups v1, and installing docker on a machine with LabVIEW Real Time 2025 results in mounting both v1 and v2 paths. Eliminating this dependency from docker enables running simple test containers on NILRT.

### Justification

[AB#3271314](https://dev.azure.com/ni/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/3271314)


### Testing

I have tested the following
1. cgroup v2 paths are mounted when LabVIEW is not installed
2. cgroup v1 paths are mounted when LabVIEW 2023 is installed
3. cgroup v2 paths are mounted when LabVIEW 2025 is installed
4. I was able to run `docker run hello-world` with cgroups v2 - I had to remove cgroups-lite as a dependency of docker-moby because that's for cgroups v1 specificially

This will be followed up by more testing as specified in the [spec](https://dev.azure.com/ni/DevCentral/_git/labview?path=%2Fdocs%2FLabVIEW-wiki%2FSpecifications%2FDeveloping-Features%2FUse-cgroups-v2-in-LabVIEW-RT%2FUse-cgroups-v2-in-LabVIEW-RT.md&_a=preview).

* [x] I have built the core package feed with this PR in place. (`bitbake packagefeed-ni-core`)


### Procedure

* [x] I certify that the contents of this pull request complies with the [Developer Certificate of Origin](https://github.com/ni/nilrt/blob/HEAD/docs/CONTRIBUTING.md#developer-certificate-of-origin-dco).

gratian commented on recipes-ni/initscripts-nilrt/files/nicreatecpuacctgroups:
```diff
@@ -3,40 +3,64 @@
+	mkdir other_group
+
+	# fast_task_writes (v1 only)
+	[ -e ./other_group/cgroup.fast_task_writes ] && echo 1 > ./other_group/cgroup.fast_task_writes
```
We can drop all references to `cgroup.fast_task_writes`. As one of the comments which got deleted below states 

_"fast_task_writes was added to 3.2 for performance reasons, but does not exist and is not necessary in the 3.10 kernel."_

We're now way past the 3.10 kernel. We don't have to carry this around anymore.

pratheekshasn replied:
I have removed references to `fast_tasks_writes`.

gratian commented on recipes-ni/initscripts-nilrt/files/lvrt-cgroup.sh:
```diff
@@ -7,15 +7,24 @@ identify_lvrt_cgroup_version()
 	[ -z "$LVRT_CGROUP_VERSION" ] || return 0
 
-	# If lvrt is not installed set version to 0, cgroups not used
+	# If lvrt is not installed set version to 2
```
Maybe we can make this comment a bit more explicit to make it more useful. Something like "If LabVIEW RT is not installed default to cgroups v2"

pratheekshasn replied:
I have rephrased the comment.

gratian commented on recipes-ni/initscripts-nilrt/files/nicreatecpuacctgroups:
```diff
@@ -3,40 +3,64 @@
+	while read TASK; do
+		echo $TASK > "./other_group/tasks" 2> /dev/null
+	done < "tasks"
+	[ -e ./other_group/cgroup.fast_task_writes ] && echo 0 > ./other_group/cgroup.fast_task_writes
```
Same as above, this dead code now. Let's remove the writes to `cgroup.fast_task_writes`

pratheekshasn replied:
I have removed references to `fast_tasks_writes`.

gratian commented on recipes-ni/initscripts-nilrt/files/nicreatecpuacctgroups:
```diff
@@ -3,40 +3,64 @@
-	# tasks can be ignored in this loop and the following loop. 
-done < "tasks"
-[ -e ./other_group/cgroup.fast_task_writes ] && echo 0 > ./other_group/cgroup.fast_task_writes
+	[ -e ./ISR_group/cgroup.fast_task_writes ] && echo 1 > ./ISR_group/cgroup.fast_task_writes
```
Same as above, remove.

pratheekshasn replied:
I have removed references to `fast_tasks_writes`.

gratian commented on recipes-ni/initscripts-nilrt/files/nicreatecpuacctgroups:
```diff
@@ -3,40 +3,64 @@
+	while read TASK; do
+		echo $TASK > "./ISR_group/tasks" 2> /dev/null
+	done < "/dev/cgroup/cpuset/tasks"
+	[ -e ./ISR_group/cgroup.fast_task_writes ] && echo 0 > ./ISR_group/cgroup.fast_task_writes
```
Ditto, remove.

pratheekshasn replied:
I have removed references to `fast_tasks_writes`.

gratian commented on recipes-ni/initscripts-nilrt/files/nicreatecpusets:
```diff
@@ -3,42 +3,69 @@
+	cat mems > ./LabVIEW_tl_set/mems
+
+	cat cpus > ./LabVIEW_ScanEngine_set/cpus
+	cat mems > ./LabVIEW_ScanEngine_set/mems
```
We should use the opportunity and clean/shorten the code above up using a 'for loop' like you have below for cgroups v2. The `mkdir` and `cat` calls can probably go into the same loop.

pratheekshasn replied:
I have cleaned up the code and added a for loop.

gratian commented on recipes-ni/initscripts-nilrt/files/nicreatecpusets:
```diff
@@ -3,42 +3,69 @@
+	cat mems > ./LabVIEW_ScanEngine_set/mems
+
+	# fast_task_writes (v1 only)
+	[ -e ./system_set/cgroup.fast_task_writes ] && echo 1 > ./system_set/cgroup.fast_task_writes
```
Let's drop this. It's not in use anymore.

pratheekshasn replied:
I have removed references to `fast_tasks_writes`.

gratian commented on recipes-ni/initscripts-nilrt/files/nicreatecpusets:
```diff
@@ -3,42 +3,69 @@
+	while read TASK; do
+		echo $TASK > "./system_set/tasks" 2> /dev/null
+	done < "tasks"
+	[ -e ./system_set/cgroup.fast_task_writes ] && echo 0 > ./system_set/cgroup.fast_task_writes
```
Drop.

pratheekshasn replied:
I have removed references to `fast_tasks_writes`.

gratian reviewed (commented):
I did a first quick pass. I will take a closer look but wanted to post the comments I've had so far.

pratheekshasn committed:
initscripts-nilrt: Support cgroups v1 and v2 based on installed LVRT version

For targets with LVRT versions 2023 and lesser, mount cgroups v1 paths.
For LVRT versions 2024 and greater, mount cgroups v2 paths.
For targets with no LVRT, default to cgroups v2 paths.

Signed-off-by: pratheekshasn <pratheeksha.s.n@ni.com>

pratheekshasn commented:
Decided to change the approach so that there's a new recipe that takes care of installing these initscripts. See PR #890.
